service/geopmdpy/session.py

```diff
@@ -32,90 +32,237 @@
+        return result
+
+    def format_double(self, signal):
+        return '{:.16g}'.format(signal)
+
+    def format_integer(self, signal):
+        return '{:d}'.format(int(signal))
+
+    def format_hex(self, signal):
+        return '0x{:08x}'.format(int(signal))
+
+    def format_raw64(self, signal):
+        return '0x{:016x}'.format(int(signal))
+
+    def num_sample(self, duration, period):
+        """Number of samples required for reading
+
+        The user specifies a duration and period, this method
+        interprets how many samples should be read to satisfy the
+        request.
+
+        Args:
+            duration (float): The user specified length of time for
+                              the samples to span in units of seconds.
+            period (float): The user specified period between samples
+                            in units of seconds.
+
+        Returns:
+            int: The required number of samples
+
+        """
+        if period == 0:
+            result = 1
+        else:
+            result = int(duration / period)  + 1
+        return result
+
+    def timed_loop_init(self, period):
+        self._target_time = time.time() + period
+        self._period = period
+
+    def timed_loop_step(self):
+        sleep_time = self._target_time - time.time()
+        if sleep_time > 0:
+            time.sleep(sleep_time)
+        self._target_time += self._period
+
+    def run_read(self, duration, period, out_stream):
+        self._geopm_proxy.PlatformOpenSession()
+        num_sample = self.num_sample(duration, period)
+        self.timed_loop_init(period)
+        for sample_idx in range(num_sample):
+            out_stream.write(
+                self.format_signals(
+                    self.read_signals()))
+            if sample_idx != num_sample - 1:
+                self.timed_loop_step()
+        self._geopm_proxy.PlatformCloseSession()
+
+    def run_write(self, duration):
+        self._geopm_proxy.PlatformOpenSession()
+        for rr in self._requests:
+            self._geopm_proxy.PlatformWriteControl(*rr)
+        time.sleep(duration)
+        self._geopm_proxy.PlatformCloseSession()
+
+    def iterate_stream(self, request_stream):
+        for line in request_stream:
+            line = line.strip()
+            if line == '':
+                break
+            if not line.startswith('#'):
+                yield line
+
+    def parse_read_requests(self, request_stream):
+        self._requests = []
+        for line in self.iterate_stream(request_stream):
+            words = line.split()
+            if len(words) != 3:
+                raise RuntimeError('Invalid command for reading: "{}"'.format(line))
+            signal_name = words[0]
+            domain_type = topo.domain_type(words[1])
+            domain_idx = int(words[2])
+            self._requests.append((signal_name, domain_type, domain_idx))
+
+    def init_signal_format(self):
```

Comment:
Some docstrings would help explain what's going on here.

Why is the 0th request (of potentially multiple?) being paried with the 4th value back from SignalInfo?  I assume this is to extract the format function?

What if there are multiple requests?

Reply:
each request is a tuple (signal_name, domain_type, domain_idx).  Later we use the signal_name (the 0th element of the request) to get back an info tuple.  Element 4 of the info tuple is the format enum.